sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClient.java

```diff
@@ -726,28 +725,16 @@ Flux<ServiceBusReceivedMessage> peekMessages(int maxMessages, String sessionId)
                 final long nextSequenceNumber = lastPeekedSequenceNumber.get() + 1;
                 LOGGER.atVerbose().addKeyValue(SEQUENCE_NUMBER_KEY, nextSequenceNumber).log("Peek batch.");
 
-                final Flux<ServiceBusReceivedMessage> messages =
-                    node.peek(nextSequenceNumber, sessionId, getLinkName(sessionId), maxMessages);
-
-                // To prevent it from throwing NoSuchElementException in .last(), we produce an empty message with
-                // the same sequence number.
-                final Mono<ServiceBusReceivedMessage> handle = messages
-                    .switchIfEmpty(Mono.fromCallable(() -> {
-                        ServiceBusReceivedMessage emptyMessage = new ServiceBusReceivedMessage(BinaryData
-                            .fromBytes(new byte[0]));
-                        emptyMessage.setSequenceNumber(lastPeekedSequenceNumber.get());
-                        return emptyMessage;
-                    }))
-                    .last()
-                    .handle((last, sink) -> {
+                return node
+                    .peek(nextSequenceNumber, sessionId, getLinkName(sessionId), maxMessages)
+                    .doOnNext(next -> {
                         final long current = lastPeekedSequenceNumber
-                            .updateAndGet(value -> Math.max(value, last.getSequenceNumber()));
+                            .updateAndGet(value -> Math.max(value, next.getSequenceNumber()));
 
-                        LOGGER.atVerbose().addKeyValue(SEQUENCE_NUMBER_KEY, current).log("Last peeked sequence number in batch.");
-                        sink.complete();
+                        LOGGER.atVerbose()
+                            .addKeyValue(SEQUENCE_NUMBER_KEY, current)
+                            .log("Last peeked sequence number in batch.");
                     });
-
-                return Flux.merge(messages, handle);
```

Comment:
this merge causes two peek requests to be sent, but only the second request results to be returned to the user. 

E.g.
```java
AtomicInteger it = new AtomicInteger();
var flux = Flux.create(sink -> {
            int next = it.getAndIncrement();
            sink.next(next);
            System.out.println("next " + next);
            sink.complete();
        })
        .doFinally((st) -> System.out.println("finally " + st.toString()));

var handle = flux.last().handle((l, s) -> s.complete());

System.out.println("last " + Flux.merge(flux, handle).blockLast());
```

produces the following output:
```text
next 0
finally onComplete
next 1
finally cancel
last 0
```

 
Back to receiver.peek - if the first and second requests return the same set of messages - it's not efficient, but when requesting messages as they are being published, the second request may contain new messages which are dropped. As a result, `lastPeekedSequenceNumber` is updated and peek requests on the same client will never return dropped messages.

Reply:
nice!!